### DISCUSSION TITLE
PolymorphismStyle.None resulting in an error

### DISCUSSION THREAD

**Discussion by PhilippNowak96** [charleskorn/kaml#514] [Q&A] [answered]
I saw that `PolymorphismStyle.None` has been released recently and wanted to try it out for some usecase in my project.

For the parsing process I have the following example model class:

```kotlin
@Serializable
data class Specification(
    val params: List<Param>,
)

@Serializable
sealed class Param {
    abstract val name: String
    @SerialName("desc") abstract val description: String
    abstract val valueType: String
}
```

Then I try to parse it with the following example:

```kotlin
Yaml(
    configuration = YamlConfiguration(
        allowAnchorsAndAliases = true,
        polymorphismStyle = PolymorphismStyle.None,
    )
).decodeFromString(
    deserializer = Specification.serializer(),
    string = """
        |params:
        |  - &param_EventStep
        |    name: event_step
        |    desc: "Something"
        |    valueType: string
    """.trimMargin(),
)
```

Unfortunately I receive an error which looks like:

```
Exception in thread "main" IncorrectTypeException at params[0] on line 2, column 5: Encountered a polymorphic map descriptor but PolymorphismStyle is 'None'
	at com.charleskorn.kaml.YamlInput$Companion.createFor$kaml(YamlInput.kt:67)
	at com.charleskorn.kaml.YamlListInput.decodeElementIndex(YamlListInput.kt:38)
	at kotlinx.serialization.internal.AbstractCollectionSerializer.merge(CollectionSerializers.kt:34)
	at kotlinx.serialization.internal.AbstractCollectionSerializer.deserialize(CollectionSerializers.kt:43)
	at kotlinx.serialization.encoding.Decoder$DefaultImpls.decodeSerializableValue(Decoding.kt:257)
	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:16)
```

Can you tell me what exactly the problem is? As far as I understood the new `polymorphismStyle = PolymorphismStyle.None` should just skip the checks regarding polymorphism, i.e. doesn't care about sealed class and stuff like that. Am I wrong?

**Comment by charleskorn** [comment, marked as answer]
`PolymorphismStyle.None` does not support deserializing at all - it is for serialization only.

In your case, I'd suggest creating a custom serializer for the `Param` type that constructs the correct subtype of `Param` based on the value of `valueType`. There's more information about this in the kotlinx.serialization docs [here](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/serializers.md#custom-serializers). (Note that the examples refer to JSON, but the same ideas apply for YAML.)

**Reply by PhilippNowak96** [reply]
Oh then I misunderstood that mode 😅 Thank you very much for the explanation, I will definitely try the custom serializer way.

**Comment by mgroth0** [comment]
> PolymorphismStyle.None does not support deserializing at all - it is for serialization only.

Exactly. I requested this feature because I wanted to serialize a polymorphic tree. The resulting yaml would be read by an external program that does not check types.
